Add `peek_lfu_key` methods

cyqsimon opened the pull request:
I noticed that there is no way to get the key that the evicted item (or the item about to be evicted) was/is stored under.

This PR adds a `peek_lfu_key` method for this particular need of mine.

edward-shen commented:
Thanks!

edward-shen commented:
1.3.0 is released with your changes!

edward-shen commented:
If you'd like, you can also put up these changes for the 2.0 branch. There are some merge conflicts (primarily due to the change in visibility on the functions), but fixing them removes the pgp signature from your commits.

cyqsimon commented:
> If you'd like, you can also put up these changes for the 2.0 branch. There are some merge conflicts (primarily due to the change in visibility on the functions), but fixing them removes the pgp signature from your commits.

Wilco. Although I don't really have a problem with it not having my signature 😄.

---

Also I would like to discuss with you the possibility of returning a key-value pair instead of just the value on eviction. Maybe adding a new family of functions? Or maybe changing the existing function signatures? Seeing that you are planning 2.0, I feel like this may be an appropriate time for this.

Reason for this is I'm currently implementing a file cache using your crate, which requires flushing to disk on eviction, with the filename being composed using the key. Since I didn't want to cause breaking changes, I just picked the simplest (and laziest) way to obtain this key and submitted this PR. But it does mean I end up with this kind of jank:

```rust
// when peek_lfu_key() returns `Some`, it just means there is at least 1 item;
// an eviction will not necessarily happen on the next insertion
let flush_key = self.cache.peek_lfu_key().cloned();
let evicted_item = self.cache.insert(key.clone(), item);

match (flush_key, evicted_item) {
    (Some(key), Some(evicted)) => {
        let flush_path = self.get_file_path(key);
        evicted.flush(flush_path).await?;
    }
    (_, None) => {
        // nothing evicted
    }
    (None, Some(_)) => unreachable!("something is wrong with the LFU implementation"),
};
```

It would be much more ideal if `cache.insert` could return both the key and value instead.

edward-shen commented:
To be clear, `2.0` has had very slow progress. I don't expect it to be released for another year or two!

The reason why that function returns a `Option<V>` instead of an `Option<(K, V)>` is just to  match `std::collections::HashMap` semantics. I'm trying to stay close to matching the signatures of those methods as close as possible; `HashMap::insert` returns a `Option<V>`.

Would the Entry API work here?

cyqsimon commented:
> The reason why that function returns a `Option<V>` instead of an `Option<(K, V)>` is just to match `std::collections::HashMap` semantics. I'm trying to stay close to matching the signatures of those methods as close as possible;

This is another issue I want to talk about. Honestly, I don't think this is a great idea. Reason is `HashMap` does not have the concept of "eviction". When `HashMap::insert` returns a `Some`, I can be sure that it's because the key already existed prior to the insertion, and that the returned value is the old value.

However when you add in the concept of eviction, returning a single `Option` becomes inadequate. It is no longer sufficient to encapsulate the two semantically-different outcomes expressively (i.e. did I receive a `Some` because the key was already present? Or was it because something was evicted?). There are plenty of cases where these two require different handling, so this feels like a mistake that's asking to be made.

In my opinion, this kind of situation calls for a new enum, something like this:

```rust
pub enum InsertResult<K, V> { // Ugh, I don't like this name honestly. Naming is hard =(.
  Added, // unseen key, no evictions
  Replaced(V), // seen key, old value replaced
  Evicted((K, V)), // unseen key, unrelated key-value pair evicted
}
```
